integration_tests/src/fixture_generator/parser.rs

```diff
@@ -0,0 +1,153 @@
+    /// The following are valid examples of user arguments:
+    ///   --user "username: 1, 3, (5, 'xyz')"
+    ///   --user="username: [(1,'abc'), 2]"
+    ///   -u "some username: 1"
+    #[clap(
+        long = "user",
+        short = 'u',
+        multiple_occurrences = true,
+        parse(try_from_str = parse_user_events),
+    )]
+    pub users: Vec<User>,
+
+    /// Number of epochs to advance the tree by
+    /// e.g. a value of 3 will perform 3 publishes on an empty directory.
+    #[clap(long = "epochs", short = 'e')]
+    pub epochs: u32,
+
+    /// Maximum number of updates **per epoch** the tool should perform.
+    /// Note that all user events explicitly passed for an epoch will be
+    /// included even if the number exceeds this value.
+    #[clap(long = "max_updates", default_value = "10")]
+    pub max_updates: u32,
+
+    /// Minimum number of updates **per epoch** the tool should perform.
+    /// The tool will generate random labels and values to include in an epoch
+    /// if the user events explicitly passed for an epoch are not sufficients.
+    #[clap(long = "min_updates", default_value = "0")]
+    pub min_updates: u32,
+
+    /// Epochs where the state of the directory should be captured in the output
+    /// e.g. the value 3 will output all db records after epoch 3 is performed.
+    /// Multiple values should be separated with ","
+    #[clap(
+        long = "capture_states",
+        short = 's',
+        multiple_values = true,
+        use_value_delimiter = true,
+        require_value_delimiter = true
+    )]
+    pub capture_states: Option<Vec<u32>>,
+
+    /// Epochs where the updates required to bring the directory to the epoch
+    /// should be captured in the output.
+    /// e.g. the value 3 will output all updates that were performed to advance
+    /// the directory from epoch 2 to 3.
+    /// Multiple values should be separated with ","
+    #[clap(
+        long = "capture_deltas",
+        short = 'd',
+        multiple_values = true,
+        use_value_delimiter = true,
+        require_value_delimiter = true
+    )]
+    pub capture_deltas: Option<Vec<u32>>,
+
+    /// Name of output file.
+    /// If omitted, output will be printed to stdout.
+    #[clap(long = "out", short = 'o')]
+    pub out: Option<String>,
+
+    /// Stops tool from generating random updates in publishes.
+    /// Use this if you want the tool to only use explicitly passed updates.
+    /// Explicilty passed updates without values would still use randomly
+    /// generated values.
+    #[clap(long = "no_generated_updates", short = 'n')]
+    pub no_generated_updates: bool,
+}
+
+fn parse_user_events(s: &str) -> Result<User, String> {
+    let mut split = s.split(':');
+    let username_text = split.next().unwrap();
+    let maybe_events_text = split.next();
+
+    let username = Regex::new(USER_PATTERN)
+        .unwrap()
+        .captures(username_text)
+        .unwrap()
+        .get(0)
+        .unwrap()
```

Comment:
nit: There's a lot of unwraps here, which will panic on invalid input right? Is it better to do the error propogation which will fail the parser but print a cleaner message to the operator about what went wrong?

Reply:
Since we're not generating the Regex dynamically, it's probably safe enough on ```Regex::new(<pattern>).unwrap()```

Reply:
Ah I see you called this out up front in the PR notes, so just as long as we track it with an issue, I'm fine with it for now.